### PR TITLE
pm: System Power Manager performs devices power transition on specified Core

### PR DESCRIPTION
**Pull request by aborisovich**
Added new feature to Device Power Management Infrastructure that allows to define a new option CONFIG_PM_DEVICE_ACTION_CORE_ID. This option defines ID of the Core that will perform devices power transition (suspend/resume).
When multiple cores enter `pm_system_suspend()` and enable `CONFIG_PM_DEVICE_ACTION_CORE_ID` function will simply return and will not execute suspend.
This was originally done as fix to the issue with `z_cpus_active` that is resolved by the https://github.com/zephyrproject-rtos/zephyr/pull/58446 .
However, I think this approach may be beneficial as new feature, because:

- Zephyr application may wish to execute [pm_device_action](https://docs.zephyrproject.org/latest/services/pm/api/index.html#c.pm_device_action) suspend/resume on the concrete Core.
Currently it would require adding a runtime check in every definition of device_pm_action:
`if (cpuId != expectedCpuId) return;`
Or alternatively implementing a custom PM policy to make sure only given core is active before suspend action is taken.
With this approach we have a warranty of specified core performing a Device suspend/resume.
- A check to `arch_cpu_active()` has some performance cost but we are in the idle thread so it is not that relevant.

Signed-off-by: Andrey Borisovich <andrey.borisovich@intel.com>

### CODE REVIEW

**Comment by aborisovich** [comment]
@ceolin is this acceptable or we should not block the idle thread?

**Review by ceolin — CHANGES_REQUESTED** [review]
There are some fundamental problems with this implementation:

1) You can't block the idle thread.
2) Using a semaphore here does not make sense. The idle thread runs only when there is no other thread to run, if you attempt to take a semaphore that is not available what happens ? The scheduler has no other thread to schedule. This is just a like busy_wait in an atomic variable but doing way more things.
3) That is making the logic way complicated and hard to follow.

I got your use case and it makes sense for me. It is hard to do it out-of-box on Zephyr PM subsystem, but I believe it can be done writing a custom power management policy. There you can always return ACTIVE to the core that you want to be responsible to suspend devices while other cores are still running.

**Comment by aborisovich** [comment]
@ceolin I've changed the implementation and it makes more sense now.
Removed all semaphores and blocking mechanisms.
It turned out that in SOF we enter `pm_system_suspend()` on Core0 (primary) only. All secondary cores are power down at this time. That is why, existing implementation that used `z_cpus_active` variable had to be adjusted (as it assumed that all cores are active and entering `pm_system_suspend()` function).
In terms of generic solution - this PR allows to specify the core ID on which devices will be suspended however provide no synchronization between cores suspension (what is I think impossible to do since we can not block idle thread).
So if we would have other kind of architecture that is using `pm_system_suspend()` on all active cores and enables `CONFIG_PM_DEVICE_ACTION_CORE_ID`, then the order of cores suspend/resume is undetermined/unsynchronized.

**Comment by ceolin** [comment]
> @ceolin I've changed the implementation and it makes more sense now. Removed all semaphores and blocking mechanisms. It turned out that in SOF we enter `pm_system_suspend()` on Core0 (primary) only. All secondary cores are powered down at this time. That is why, existing implementation that used `z_cpus_active` variable had to be adjusted (as it assumed that all cores are active and entering `pm_system_suspend()` function?). In terms of generic solution - this PR allows to specify the core ID on which devices will be suspended however provide no synchronization between cores suspension (what is I think impossible to do since we can not block idle thread). So if we would have other kind of architecture that is using `pm_system_suspend()` on all active cores and enables `CONFIG_PM_DEVICE_ACTION_CORE_ID`, then the order of cores suspend/resume is undetermined/unsynchronized.

That is better indeed, still (as you mentioned) this solution does not ensure the order that cores are suspended and you may have core0 suspending all devices and itself while other cores are still active. This seems wrong. You can synchronize the order that cores are suspended creating a custom policy.

**Comment by aborisovich** [comment]
> > @ceolin I've changed the implementation and it makes more sense now. Removed all semaphores and blocking mechanisms. It turned out that in SOF we enter `pm_system_suspend()` on Core0 (primary) only. All secondary cores are powered down at this time. That is why, existing implementation that used `z_cpus_active` variable had to be adjusted (as it assumed that all cores are active and entering `pm_system_suspend()` function?). In terms of generic solution - this PR allows to specify the core ID on which devices will be suspended however provide no synchronization between cores suspension (what is I think impossible to do since we can not block idle thread). So if we would have other kind of architecture that is using `pm_system_suspend()` on all active cores and enables `CONFIG_PM_DEVICE_ACTION_CORE_ID`, then the order of cores suspend/resume is undetermined/unsynchronized.
> 
> That is better indeed, still (as you mentioned) this solution does not ensure the order that cores are suspended and you may have core0 suspending all devices and itself while other cores are still active. This seems wrong. You can synchronize the order that cores are suspended creating a custom policy.

This is true, however I think I have another improvement: we can introduce this feature with well documented assumption, that cores other than `CONFIG_PM_DEVICE_ACTION_CORE_ID` need to be powered down so this power transition works.
What I mean is simply return from `pm_system_suspend()` function if the remaining cores are not in the powered down state.

**Comment by aborisovich** [comment]
Push update:

- resolved matter with undetermined order of multiple cores during pm_system_suspend - took approach to require all other cores to be powered down or the pm_system_suspend will simply return and not take effect.
- added check with architecture specific function `bool arch_cpu_active(int id)` to check whether all cores but the one with `CONFIG_PM_DEVICE_ACTION_CORE_ID` are powered down.

@ceolin will this do? I think this implementation makes sense. Now we take an alternative approach for the System Device API - we do not demand from the Zephyr application to implement a PM policy. Now application may call `pm_state_force()` to bypass policy, shut down all the cores it does not need and we do the Device management and trigger PM notifiers on the single well known core. The only requirement is that given architecture needs to implement its `arch_cpu_active()` function.
@nashif are there any concerns to such approach? To demand architecture specific function implementation?
In my opinion it is logical - if you wish to use this behavior, define how to check for core state.

**Comment by aborisovich** [comment]
@ceolin is there any chance it gets some attention or should I just close it?

**Comment by ceolin** [comment]
> @ceolin is there any chance it gets some attention or should I just close it? I mean you fixed the related defect so this PR is just an idea/proposition of how alternatively pm might work. If you don't like it or see that is does not match Zephyr architecture or something please say it and I will close it.

I think this should be closed, simply because it still requires that the desired core is the last one suspended  and if that is the case you don't need any additional logic.
